L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc

```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
-  // As these inherit from L1Candidate, could set their HW quantities to the bit accurate ones
-  for (const auto& rejected_cluster : rejected_clusters_out) {
-    rejected_clusters.push_back(0, *clustersPtrs.at(rejected_cluster.index_cmssw()));
-  }
+    l1thgcfirmware::HGCalClusterSAPtrCollection& clusterSums,
+    l1t::HGCalMulticlusterBxCollection& multiClusters_out,
+    const std::vector<std::vector<edm::Ptr<l1t::HGCalCluster>>>& inputClustersPtrs) const {
+  for (const auto& cluster : clusterSums) {
+    // Convert from digitised quantities
+    if (cluster->w() == 0 || cluster->e() == 0)
+      continue;
+    double phi = (cluster->wphi() / cluster->w()) * theConfiguration_.phiRange() / theConfiguration_.phiNValues();
+    double pt = cluster->e() / theConfiguration_.ptDigiFactor();
 
-  std::vector<l1t::HGCalMulticluster> multiclusters;
-  multiclusters.reserve(multiclusters_out.size());
-  for (unsigned int imulticluster = 0; imulticluster < multiclusters_out.size(); ++imulticluster) {
-    bool firstConstituent = true;
-    for (const auto& constituent : multiclusters_out[imulticluster].constituents()) {
-      if (firstConstituent) {
-        multiclusters.emplace_back(clustersPtrs.at(constituent.index_cmssw()), 1.);
-      } else {
-        multiclusters.at(imulticluster).addConstituent(clustersPtrs.at(constituent.index_cmssw()), 1.);
-      }
-      firstConstituent = false;
+    if (pt < 0.5)
+      continue;  // Add (or take) cut threshold to config
+
+    double rOverZ =
+        (cluster->wroz() / cluster->w()) * theConfiguration_.rOverZRange() / theConfiguration_.rOverZNValues();
+    double eta = -1.0 * std::log(tan(atan(rOverZ) / 2));
+    eta *= theConfiguration_.zSide();
+
+    auto sector = theConfiguration_.sector();
+    if (sector == 1) {
+      phi += (2. * M_PI / 3.);
+    } else if (sector == 2) {
+      phi = phi + (4. * M_PI / 3.);
     }
-  }
+    if (theConfiguration_.zSide() == 1) {
+      phi = M_PI - phi;
+    }
+    phi -= (phi > M_PI) ? 2 * M_PI : 0;
+
+    math::PtEtaPhiMLorentzVector clusterP4(pt, eta, phi, 0.);
+
+    l1t::HGCalMulticluster multicluster;
+    multicluster.setP4(clusterP4);
+    std::cout << "Got a cluster : " << cluster->e() << " " << cluster->constituents().size() << " " << multicluster.pt()
+              << " " << multicluster.eta() << " " << multicluster.phi() << " " << rOverZ << " " << phi << " " << eta
+              << std::endl;
+    for (const auto& tc : cluster->constituents()) {
+      const auto& tc_cmssw = inputClustersPtrs.at(tc->cmsswIndex().first).at(tc->cmsswIndex().second);
+      // Add tc as constituent, but don't update any other properties of the multicluster i.e. leave them unchanged from those calculated by the emulator
+      multicluster.addConstituent(tc_cmssw, false, 0.);
+    }
+
+    double emIntFraction = 1.0 * cluster->e_em() / cluster->e();
```

Comment:
Why is the `1.0*` needed?
If this is for conversion into float, then it would be better to have an explicit conversion

Reply:
Using explicit conversion to float.